File: HotelServer/src/main/java/com/hotel/entity/Hotel.java
package com.hotel.entity;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Hotel {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;
	private String name;
	private String location;
	private int rating;
	private String description; // Brief description of the hotel
    private String contact; // Hotel contact number
    private String email;

	@ManyToOne
	@JoinColumn(name = "user_id", nullable = false)
	private User user; // Link hotel to a user
	@JsonManagedReference
	@OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
	private List<Room> rooms; // One-to-Many relationship with Room

	@OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
	private List<HotelImage> hotelImages; // Ensures images are deleted with the hotel

	@OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
	private List<RoomImage> roomImages;
}

File: Hotel-Client/src/app/models/Hotel.ts
import { HotelImage } from "./HotelImage";
import { RoomImage } from "./RoomImage";
import { Room } from "./Rooms";
import { User } from "./User";

export class Hotel {
  id?: number; // Optional for new hotels
  name: string;
  location: string;
  rating: number;
  description: string;
  contact: string;
  email: string;
  user?: User; // Optional reference to User model
  rooms?: Room[]; // List of rooms (optional)
  hotelImages?: HotelImage[]; // List of hotel images (optional)
  roomImages?:RoomImage[];
  standardRoomPrice?: number | null; 

  constructor(
    name: string = '',
    location: string = '',
    email: string = '',
    description: string = '',
    contact: string = '',
    rating: number = 0,
    user?: User, // Accepts an optional User object
    rooms: Room[] = [],
    hotelImages: HotelImage[] = [],
    roomImages:RoomImage[]=[],
    
  ) {
    this.name = name;
    this.description = description;
    this.contact = contact;
    this.email = email;
    this.location = location;
    this.rating = rating;
    this.user = user; // Now accepts a User object instead of just an ID
    this.rooms = rooms;
    this.hotelImages = hotelImages;
    this.roomImages=roomImages;
  }
}
